src/System.Net.Utilities/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -2,33 +2,268 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System.ComponentModel;
+using System.Net.Sockets;
+using System.Runtime.InteropServices;
 using System.Threading;
+using System.Diagnostics;
+using System.Text;
+using System.IO;
 
 namespace System.Net.NetworkInformation
 {
     public partial class Ping
     {
-        private void InternalSendAsync(IPAddress address, byte[] buffer, int timeout, PingOptions options)
+        private const int IcmpHeaderLengthInBytes = 8;
+        private const int IpHeaderLengthInBytes = 20;
+
+        private static readonly bool s_canUseRawSockets = CheckRawSocketPermissions();
+
+        // Ubuntu has ping under /bin, OSX under /sbin.
+        private static readonly string[] s_binFolders = { "/bin", "/sbin" };
```

Comment:
I'm a bit surprised any of this manual path manipulation stuff is necessary, and I also worry it's going to change from system to system.  Process.Start already looks in a variety of places if the supplied path isn't rooted (https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L298), including checking every directory in the PATH environment variable, which should include both bin and sbin.